### DISCUSSION TITLE
Use foreman as a provider for broker

### DISCUSSION THREAD

**Discussion by dosas** [SatelliteQE/broker#284] [Ideas]
Would you be open to a PR to use foreman as an additional provider?

**Comment by JacobCallahan** [comment]
@dosas that's an interesting idea that I've floated around in the past. As we know there is a large amount of scope that this could encompass, but if we were to add foreman it would likely be assumed that the entire provisioning setup has been done prior to Broker being called. Does this fit in line with what you were thinking?

**Reply by knoppi** [reply]
Basically yes. (I'm answering as @dosas partner in crime).

We have a Foreman A which has been running for quite a long time and broker is used within robottelo tests targetting Foreman B. Workflow is roughly this:

1) Provision EL8 host using Foreman A
2) Install and configure Foreman B on this host
3) Run robottelo tests.

So, during robottelo tests we still have access to Foreman A and hosts deployed with it are easily accessible regarding network routes, DNS, SSH keys, ... Moreover you can prepare Foreman A to deploy further hosts with only very few parameters (hostgroup, image and compute resource is what we are thinking currently). That's why we naturally thought about using it.

**Reply by JacobCallahan** [reply]
Checkout/Checkin should be pretty standard with a foreman provider. However, would we want to offer Execute in the form of remote execution?
Also, we'll need an interaction layer. Would that be through nailgun or do you already have something else in mind?

**Reply by knoppi** [reply]
Hey, I already implemented checkout/checkin and indeed I already have a method to execute via remote execution. That basically works well, but I had to do an ugly hack to determine the hostname.

I currently wrote everything as pure API queries (python requests). Would you use nailgun for these operation or is interaction layer something else?

**Reply by ogajduse** [reply]
If I were implementing the Foreman provider, I would give https://github.com/Apipie/apypie a shot.
Foreman devs use it in the foreman-ansible-modules project.
https://github.com/theforeman/foreman-ansible-modules/blob/develop/plugins/module_utils/foreman_helper.py

They were using nailgun in this project before, but since nailgun was too heavy for them, they switched to apypie.

**Comment by knoppi** [comment]
I've opened a PR for our broker's draft: https://github.com/SatelliteQE/broker/pull/291
